File: src/native/printer-plugin/registry/connection-registry.test.ts
import { describe, it, expect } from 'vitest'
import { ConnectionRegistry } from './connection-registry'
import type { PrinterConnection } from '../internal/types'
import type { PrinterDevice } from '../definitions'

class FakeConnection implements PrinterConnection {
  async connectTo(_device: PrinterDevice): Promise<void> {}
  async disconnect(): Promise<void> {}
  async write(_data: Uint8Array): Promise<void> {}
  isConnected(): boolean {
    return false
  }
}

describe('ConnectionRegistry', () => {
  it('resolves a registered connection by type', () => {
    const registry = new ConnectionRegistry()
    registry.register('usb', () => new FakeConnection())

    expect(registry.resolve('usb')).toBeInstanceOf(FakeConnection)
  })

  it('throws a clear error when the type is not registered', () => {
    const registry = new ConnectionRegistry()
    expect(() => registry.resolve('lan')).toThrow('Connection "lan" chưa được đăng ký.')
  })
})

File: android/app/src/test/java/com/ndtcore/pos/printer/registry/ConnectionRegistryTest.kt
package com.ndtcore.pos.printer.registry

import com.ndtcore.pos.printer.connection.ConnectionTarget
import com.ndtcore.pos.printer.connection.PrinterConnection
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

private class FakeConnection : PrinterConnection {
    override fun connect(target: ConnectionTarget) {}
    override fun disconnect() {}
    override fun write(data: ByteArray) {}
    override fun isConnected(): Boolean = false
}

class ConnectionRegistryTest {
    @Test
    fun `resolve returns a registered connection`() {
        val registry = ConnectionRegistry()
        registry.register("usb") { FakeConnection() }

        assertTrue(registry.resolve("usb") is FakeConnection)
    }

    @Test
    fun `resolve throws a clear error when the type is not registered`() {
        val registry = ConnectionRegistry()
        val error = assertThrows(IllegalArgumentException::class.java) { registry.resolve("lan") }
        assertTrue(error.message!!.contains("lan"))
    }
}
